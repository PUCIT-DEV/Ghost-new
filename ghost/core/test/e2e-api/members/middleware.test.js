const {agentProvider, mockManager, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyEtag, anyObjectId, anyUuid, anyISODateTime} = matchers;
const models = require('../../../core/server/models');
const should = require('should');

let membersAgent;

const memberMatcher = (newslettersCount) => {
    return {
        uuid: anyUuid,
        // @NOTE: check if this field is even needed? it differs to the output in the other matcher
        created_at: anyISODateTime,
        newsletters: new Array(newslettersCount).fill(
            {
                id: anyObjectId
            }
        )
    };
};

const buildMemberMatcher = (newslettersCount) => {
    return {
        uuid: anyUuid,
        newsletters: new Array(newslettersCount).fill(
            {
                id: anyObjectId
            }
        )
    };
};

describe('Comments API', function () {
    before(async function () {
        membersAgent = await agentProvider.getMembersAPIAgent();

        await fixtureManager.init('newsletters', 'members:newsletters');
    });

    beforeEach(function () {
        mockManager.mockMail();
    });

    afterEach(function () {
        mockManager.restore();
    });

    describe('when not authenticated but enabled', function () {
        it('can not get member data', async function () {
            await membersAgent
                .get(`/api/member/`)
                .expectStatus(204)
                .expectEmptyBody();
        });

        it('can update comment notifications', async function () {
            // Only via updateMemberNewsletters
            let member = await models.Member.findOne({id: fixtureManager.get('members', 0).id}, {require: true});
            member.get('enable_comment_notifications').should.eql(true, 'This test requires the initial value to be true');

            await membersAgent
                .put(`/api/member/newsletters/?uuid=${member.get('uuid')}`)
                .body({
                    enable_comment_notifications: false
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(buildMemberMatcher(1))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.enable_comment_notifications.should.eql(false);
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('enable_comment_notifications').should.eql(false);
        });
    });

    describe('when authenticated', function () {
        let member;

        before(async function () {
            await membersAgent.loginAs('member@example.com');
            member = await models.Member.findOne({email: 'member@example.com'}, {require: true});
        });

        it('can get member data', async function () {
            await membersAgent
                .get(`/api/member/`)
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(memberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                });
        });

        it('can update member expertise', async function () {
            await membersAgent
                .put(`/api/member/`)
                .body({
                    expertise: 'Head of Testing'
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(memberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.expertise.should.eql('Head of Testing');
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('expertise').should.eql('Head of Testing');
        });

        it('trims whitespace from expertise', async function () {
            await membersAgent
                .put(`/api/member/`)
                .body({
                    expertise: '  test  '
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(memberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.expertise.should.eql('test');
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('expertise').should.eql('test');
        });

        it('can update name', async function () {
            await membersAgent
                .put(`/api/member/`)
                .body({
                    name: 'Test User'
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(memberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.name.should.eql('Test User');
                    body.firstname.should.eql('Test');
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('name').should.eql('Test User');
        });

        it('can update comment notifications', async function () {
            member.get('enable_comment_notifications').should.eql(true, 'This test requires the initial value to be true');

            // Via general way
            await membersAgent
                .put(`/api/member/`)
                .body({
                    enable_comment_notifications: false
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(memberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.enable_comment_notifications.should.eql(false);
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('enable_comment_notifications').should.eql(false);

            // Via updateMemberNewsletters
            await membersAgent
                .put(`/api/member/newsletters/?uuid=${member.get('uuid')}`)
                .body({
                    enable_comment_notifications: true
                })
                .expectStatus(200)
                .matchHeaderSnapshot({
                    etag: anyEtag
                })
                .matchBodySnapshot(buildMemberMatcher(2))
                .expect(({body}) => {
                    body.email.should.eql(member.get('email'));
                    body.enable_comment_notifications.should.eql(true);
                });
            member = await models.Member.findOne({id: member.id}, {require: true});
            member.get('enable_comment_notifications').should.eql(true);
        });

        it('can remove a member\'s email from the suppression list', async function () {
            // add member's email to the suppression list
            await models.Suppression.add({
                email: member.get('email'),
                reason: 'bounce'
            });

            // disable member's email
            await member.save({email_disabled: true});

            // remove suppression
            await membersAgent
                .delete(`/api/member/suppression`)
                .expectStatus(204)
                .expectEmptyBody();

            // check that member is removed from suppression list
            const suppression = await models.Suppression.findOne({email: member.get('email')});

            should(suppression).be.null();

            // check that member's email is enabled
            await member.refresh();

            should(member.get('email_disabled')).be.false();
        });
    });
});
