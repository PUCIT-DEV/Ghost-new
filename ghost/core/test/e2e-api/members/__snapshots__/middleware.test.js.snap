// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments API when authenticated can get member data 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": null,
  "firstname": null,
  "name": null,
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can get member data 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "529",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can remove member from suppression list and resubscribe to default newsletters 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": "test",
  "firstname": "Test",
  "name": "Test User",
  "newsletters": Array [],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can remove member from suppression list and resubscribe to default newsletters 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "350",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can remove member from suppression list and resubscribe to default newsletters 3: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": "test",
  "firstname": "Test",
  "name": "Test User",
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can remove member from suppression list and resubscribe to default newsletters 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "540",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can update comment notifications 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": false,
  "expertise": "test",
  "firstname": "Test",
  "name": "Test User",
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can update comment notifications 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "541",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can update comment notifications 3: [body] 1`] = `
Object {
  "email": "member@example.com",
  "enable_comment_notifications": true,
  "name": "Test User",
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "status": "free",
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can update comment notifications 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "354",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can update member expertise 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": "Head of Testing",
  "firstname": null,
  "name": null,
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can update member expertise 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "542",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated can update name 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": "test",
  "firstname": "Test",
  "name": "Test User",
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated can update name 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "540",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when authenticated trims whitespace from expertise 1: [body] 1`] = `
Object {
  "avatar_image": null,
  "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
  "email": "member@example.com",
  "email_suppression": Object {
    "info": null,
    "suppressed": false,
  },
  "enable_comment_notifications": true,
  "expertise": "test",
  "firstname": null,
  "name": null,
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Newsletter",
      "sort_order": 0,
    },
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Weekly newsletter",
      "sort_order": 2,
    },
  ],
  "paid": false,
  "subscribed": false,
  "subscriptions": Array [],
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when authenticated trims whitespace from expertise 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "531",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Comments API when not authenticated but enabled can update comment notifications 1: [body] 1`] = `
Object {
  "email": "member1@test.com",
  "enable_comment_notifications": false,
  "name": "Mr Egg",
  "newsletters": Array [
    Object {
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Daily newsletter",
      "sort_order": 1,
    },
  ],
  "status": "free",
  "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
}
`;

exports[`Comments API when not authenticated but enabled can update comment notifications 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "253",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;
